File: JavaLibrary/src/test/java/com/flat/simplex/parser/logic/block/BlockScopeTest.java
package com.flat.simplex.parser.logic.block;

import com.flat.simplex.parser.logic.Context;
import com.flat.simplex.parser.logic.error.Error;
import com.flat.simplex.support.TokenChain;
import org.junit.jupiter.api.Test;

import static com.flat.simplex.support.ContextSupport.assertErrors;
import static com.flat.simplex.support.TokenChain.readChain;

class BlockScopeTest {

    @Test
    public void blockScope() {
        TokenChain chain = readChain("{hello = true;}");

        Context context = new Context();
        BlockScope block = new BlockScope(context, null, chain.get(), null);
        block.read();

        TokenChain.assertOne("{hello = true;}", block.getTokenContent(), "Invalid body");

        assertErrors(context);
    }

    @Test
    public void blockScopeOpenBlock_Fail() {
        TokenChain chain = readChain("{hello = true;");

        Context context = new Context();
        BlockScope block = new BlockScope(context, null, chain.get(), null);
        block.read();

        TokenChain.assertOne("{hello = true;", block.getTokenContent(), "Invalid body");

        assertErrors(context, Error.missingCloser);
    }

    @Test
    public void blockScopeUnexpectedToken_Fail() {
        TokenChain chain = readChain("{hello = true;};");

        Context context = new Context();
        BlockScope block = new BlockScope(context, null, chain.get(), null);
        block.read();

        TokenChain.assertOne("{hello = true;}", block.getTokenContent(), "Invalid body");

        assertErrors(context, Error.unexpectedToken);
    }

    @Test
    public void blockScopeUnexpectedEndOfTokens_Fail() {
        TokenChain chain = readChain("");

        Context context = new Context();
        BlockScope block = new BlockScope(context, null, chain.get(), null);
        block.read();

        assertErrors(context, Error.unexpectedEndOfTokens);
    }
}

File: SimplexApp/simplex/parser/logic/block/BlockScope.test.js
const TokenChain = require("simplex/support/TokenChain.js");
const Context = require("simplex/parser/logic/Context.js");
const Error = require("simplex/parser/logic/error/Error.js");
const BlockScope = require("simplex/parser/logic/block/BlockScope.js")
const ContextSupport = require("simplex/support/ContextSupport.js");

const readChain = TokenChain.readChain;
const assertErrors = ContextSupport.assertErrors;

test('blockScope', () => {
    let chain = readChain('{hello = true;}');
    let context = new Context(chain.get());
    let block = new BlockScope(context, chain.get(), null);
    block.read();
    TokenChain.assertOne('{hello = true;}', block.getTokenContent(), 'Invalid body');
    assertErrors(context);
});
test('blockScopeOpenBlock_Fail', () => {
    let chain = readChain('{hello = true;');
    let context = new Context(chain.get());
    let block = new BlockScope(context, chain.get(), null);
    block.read();
    TokenChain.assertOne('{hello = true;', block.getTokenContent(), 'Invalid body');
    assertErrors(context, Error.missingCloser);
});
test('blockScopeUnexpectedToken_Fail', () => {
    let chain = readChain('{hello = true;};');
    let context = new Context(chain.get());
    let block = new BlockScope(context, chain.get(), null);
    block.read();
    TokenChain.assertOne('{hello = true;}', block.getTokenContent(), 'Invalid body');
    assertErrors(context, Error.unexpectedToken);
});
test('blockScopeUnexpectedEndOfTokens_Fail', () => {
    let chain = readChain('');
    let context = new Context(chain.get());
    let block = new BlockScope(context, chain.get(), null);
    block.read();
    assertErrors(context, Error.unexpectedEndOfTokens);
});